src/EditorFeatures/CSharp/Highlighting/KeywordHighlighters/AsyncAwaitHighlighter.cs

```diff
@@ -26,93 +30,90 @@ protected override bool IsHighlightableNode(SyntaxNode node)
-                        if (asyncModifier.Kind() != SyntaxKind.None)
-                        {
-                            spans.Add(asyncModifier.Span);
-                        }
-                        break;
-                    }
-                case LocalFunctionStatementSyntax localFunction:
-                    {
-                        var asyncModifier = localFunction.Modifiers.FirstOrDefault(m => m.Kind() == SyntaxKind.AsyncKeyword);
-                        if (asyncModifier.Kind() != SyntaxKind.None)
-                        {
-                            spans.Add(asyncModifier.Span);
-                        }
-                        break;
-                    }
-                case AnonymousFunctionExpressionSyntax anonymousFunction:
-                    if (anonymousFunction.AsyncKeyword.Kind() == SyntaxKind.AsyncKeyword)
-                    {
-                        spans.Add(anonymousFunction.AsyncKeyword.Span);
-                    }
-                    break;
-
-                case AwaitExpressionSyntax awaitExpression:
-                    // Note if there is already a highlight for the previous token, merge it
-                    // with this span. That way, we highlight nested awaits with a single span.
-                    var handled = false;
-                    var awaitToken = awaitExpression.AwaitKeyword;
-                    var previousToken = awaitToken.GetPreviousToken();
-                    if (!previousToken.Span.IsEmpty)
-                    {
-                        var index = spans.FindIndex(s => s.Contains(previousToken.Span));
-                        if (index >= 0)
-                        {
-                            var span = spans[index];
-                            spans[index] = TextSpan.FromBounds(span.Start, awaitToken.Span.End);
-                            handled = true;
-                        }
-                    }
+            using var pooledObject = s_stackPool.GetPooledObject();
 
-                    if (!handled)
-                    {
-                        spans.Add(awaitToken.Span);
-                    }
-                    break;
+            var stack = pooledObject.Object;
+            stack.Push(node);
 
-                case UsingStatementSyntax usingStatement:
-                    if (usingStatement.AwaitKeyword.Kind() == SyntaxKind.AwaitKeyword)
-                    {
-                        spans.Add(usingStatement.AwaitKeyword.Span);
-                    }
-                    break;
+            while (stack.Count > 0)
+            {
+                var current = stack.Pop();
+                yield return current;
 
-                case LocalDeclarationStatementSyntax localDeclaration:
-                    if (localDeclaration.AwaitKeyword.Kind() == SyntaxKind.AwaitKeyword && localDeclaration.UsingKeyword.Kind() == SyntaxKind.UsingKeyword)
+                foreach (var child in current.ChildNodes())
```

Comment:
💡 This could be improved and optimized:

* Use `ChildNodesAndTokens()` to avoid boxing enumerators. Check `IsNode` before pushing to the stack.
* Use `Reverse()` to ensure the visitation order matches the original implementation.

```suggestion
                foreach (var child in current.ChildNodesAndTokens().Reverse())
```

Reply:
>Use ChildNodesAndTokens() to avoid boxing enumerators.

Indeed.  Will do.

Reply:
> Use Reverse() to ensure the visitation order matches the original implementation.

I'm ok with not doing that... i don't believe such a ordering is necessary/desirable/observable really.

Reply:
The call to `Reverse()` ends up ensuring that highlights are returned in document order. While the current backing data structure (a balanced binary tree) does not currently optimize this case, other more efficient data structures would have better performance for this case (e.g. a builder for a B+-tree using a branching factor greater than 2).

Reply:
The comment for `Reverse()` does not need to be more complicated than "reverse the children to ensure they are eventually visited in document order".

Reply:
Done.  It's also nicer for debugging.  So i added that info.  Thanks!